### PR TITLE
HTML block: Add placeholder

### PR DESCRIPTION
**Pull request by Soean**
## Description
Adds a placeholder to the HTML block.

## Screenshots (jpeg or gifs if applicable):
**Before**
before
**After**
after

### CODE REVIEW

**Comment by youknowriad** [comment]
I like this personally, but I recall this was present and has been removed at some point. cc @jasmussen 
I don't want us to keep doing and undoing stuff :)

**Comment by swissspidy** [comment]
Yep, it was removed in 3220ce8c37c626c4e4733a23d3b5d989889a3934 / #2880.

**Comment by jasmussen** [comment]
I also do like it personally, but I also recognize there's an important accessibility reason for it.

An alternate approach could be to have a `Code:` label that is visually shown before the input field. Perhaps this is shown only when the block is selected, in the name of _we can show additional fields and controls when a block is selected_. Note we'd want this label to be _the one_ that gets read aloud by screenreaders, so we'd want to revisit the aria label in such a case.

**Comment by jasmussen** [comment]
CC: @afercia

**Comment by youknowriad** [comment]
If I remember correctly, we need a label for all of these "inputs" but the placeholder can't be the label. Though, I don't think it's an issue if we have both a label and a placeholder.

**Comment by Soean** [comment]
@jasmussen If I add the HTML block to a template, the user should see what the block is about without clicking on it. So if we use a label approach, we should show it constantly.

**Comment by karmatosed** [comment]
> This is the reason why in complex forms (meaning a form with many fields) a visible label is recommended.

What about in the case of where it's not a form, this isn't a form is it? I think that distinction is important to remember. Placeholders are not a form.

We need to be super careful using the word placeholder as it has a very particular meaning and in Gutenberg we also refer to it as the 'block default content'. Placeholders as default content, are important for a lot of experience enhancement across abilities. They are a good guide as to what the content could be and as we move into a template future, they are going to be even more important. It's important to use the right term for our instance as a placeholder is more in Gutenberg sense referring to what the image placeholder does. That's a great example of a placeholder in Gutenberg terms.

I would argue that it should not have been removed in the first place. We should have guiding text like this as it's important for the experience.

**Review by youknowriad — APPROVED** [review]

**Comment by afercia** [comment]
@karmatosed whether it's technically a form or not is not relevant. It resembles a form, in the user experience. There are editable fields. These fields are visually identified just by their placeholder, which in most of the cases in Gutenberg is technically a placeholder attribute:
`<textarea placeholder="Write code…" `

Regardless, these placeholder disappear as soon as you start writing. Once you've entered content, there's no way to distinguish the blocks visually.

There's some good literature around about placeholders and why an interface shouldn't rely on them, most of these resources are listed in the core trac ticket mentioned above. I'd recommend to go through them, they're all very enlightening especially the Nielsen Norman Group one.

As said, all this applies when there are _many_ fields.

I'm not saying they should be necessarily removed, even if their usage in Gutenberg is formally against the spec. I've mentioned that as a trade-off the accessibility team agreed on using both a label and a placeholder. However, the issue stays: they disappear 🙂 

I'd like to hear everyone's thoughts about the specific issue I've tried to highlight, specifically: in the screenshot below, is the interface helping in any way users to understand what the blocks type are?

screen shot 2018-01-12 at 17 38 24

File: blocks/library/html/index.js [hunk 1]
```diff
@@ -67,6 +67,7 @@ registerBlockType( 'core/html', {
 				className="wp-block-html"
 				key="editor"
 				value={ attributes.content }
+				placeholder={ __( 'Write HTML...' ) }
```
Comment (gziolo) [inline blocks/library/html/index.js]: It should use ellipsis character `…` instead.

**Comment by gziolo** [comment]
@Soean - can you rebase it with the latest `master` version. Are there any blockers in here?

**Comment by danielbachhuber** [comment]
> The HTML block now uses CodeMirror... so I guess the placeholder prop should be set on the block and passed down to the CodeEditor component. As per the visual result, not sure it will look so nice:

This looks pretty bad. Let's pass on this for now, and revisit when a decision is made on placeholders as a whole.

**Pull request closed without merging**